rxjava: support for multiparty submissions

garyverhaegen-da opened the pull request:
This (conceptually) small set of changes adds overrides for the `submit` family of methods that let clients pass lists of `actAs` and `readAs` parties rather than a single party.

```
CHANGELOG_BEGIN

- [Java Codegen] The LedgerClient interface in the rxjava bindings has
  been expanded with more overrides to the `submit` family of methods to
  account for the new multiparty submission feature.

  This is a compile-time breaking change for anyone who has a custom
  implementation of the LedgerClient, CommandClient or
  CommandSubmissionClient interfaces. Most Java IDEs will be able to
  geenrate stub implementations that throw some variant of
  UnsupportedOperationException or RuntimeException, which is a viable
  workaround until you start using the new multiparty submission
  feature.

  We expect most users to use the provided implementations of these
  interfaces rather than implement them themselves; for those users,
  this change is purely additive.

CHANGELOG_END
```

stefanobaghino-da commented on language-support/java/codegen/src/ledger-tests/scala/com/digitalasset/CodegenLedgerTest.scala:
```diff
@@ -164,4 +167,26 @@ class CodegenLedgerTest extends AsyncFlatSpec with Matchers with TestResourceCon
+      list(Charlie),
+      MultiParty.exerciseByKeyMPFetchOtherByKey(new da.types.Tuple2(Alice, Bob), Charlie, Bob))
+
+    true shouldBe true
```
```suggestion
    succeed
```

garyverhaegen-da replied:
Changed, thanks!

stefanobaghino-da commented on language-support/java/codegen/src/ledger-tests/scala/com/digitalasset/CodegenLedgerTest.scala:
```diff
@@ -164,4 +167,26 @@ class CodegenLedgerTest extends AsyncFlatSpec with Matchers with TestResourceCon
+    val multi = new MultiParty(Alice, Bob)
+    sendCmd(client, list(Alice, Bob), list[String](), multi.create());
+
+    val read :: _ = readActiveContracts(MultiParty.Contract.fromCreatedEvent)(client)
```
Any reason to use a pattern here? The error is probably easier to read using `.head`:

```
scala> val read :: _ = List.empty[Int]
scala.MatchError: List() (of class scala.collection.immutable.Nil$)
  ... 28 elided

scala> val read = List.empty[Int].head
java.util.NoSuchElementException: head of empty list
  at scala.collection.immutable.Nil$.head(List.scala:469)
  at scala.collection.immutable.Nil$.head(List.scala:466)
  ... 28 elided
```

This also protects from some head-scratching situation if the concrete return type of `readActiveContracts` changes:

```
scala> val read :: _ = Vector(1, 2, 3).asInstanceOf[Seq[Int]]
scala.MatchError: Vector(1, 2, 3) (of class scala.collection.immutable.Vector)
  ... 28 elided

scala> val read = Vector(1, 2, 3).asInstanceOf[Seq[Int]].head
read: Int = 1
```

cocreature replied:
Interesting, iirc in Haskell it’s the opposite. The pattern gives you a better error message than calling `head`.

garyverhaegen-da replied:
Changed.

stefanobaghino-da commented on language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/helper/ListHelper.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public final class ListHelper {
+    private ListHelper() {}
+    public static <T> List<T> list(T... args) {
+        List<T> ret = new ArrayList<>();
+        for(T arg : args){
+            ret.add(arg);
+        }
+        return ret;
+    }
+}
```
Would [this](https://docs.oracle.com/javase/8/docs/api/java/util/Arrays.html#asList-T...-) work for you? Starting from Java 9 we can also use [this](https://docs.oracle.com/javase/9/docs/api/java/util/List.html#immutable).

garyverhaegen-da replied:
I'm aware of both, but we're stuck on 1.8 and I think
```
list(party), list(),
```
reads a lot better than
```
Arrays.toList(new String[]{party}), Arrays.toList(new String[]{}),
```

Somehow it never occurred to me before right now to use `Arrays.toList` in the implementation of `list`, though. On the off chance that is what you were suggesting, I've done that now.

stefanobaghino-da replied:
Are you sure that is how the syntax would work? The documentation itself says:

> This method also provides a convenient way to create a fixed-size list initialized to contain several elements:

```java
Arrays.asList("Larry", "Moe", "Curly");
```

stefanobaghino-da replied:
By the way, since we're here, in the future you may also find [Collections.singletonList](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#singletonList-T-) and [Collections.emptyList](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#emptyList--) useful.

stefanobaghino-da reviewed (commented):
Seems largely ok. I left a few early comments. I'll have a second look tomorrow and I'd recommend heavy a pair of extra eyeballs on this one. There's _a lot_ of code.

stefanobaghino-da reviewed (approved):
Looks ok, but having a couple of extra eyeballs on this won't hurt.

Furthermore, I believe we should consider using a different approach next time we need to add something, this is becoming unwieldy. Maybe using a type wrapping command submission parameters coupled with a builder can be enough.

cocreature commented on language-support/java/bindings-rxjava/src/test/scala/com/daml/ledger/rxjava/components/BotTest.scala:
```diff
@@ -151,6 +152,19 @@ final class BotTest extends AnyFlatSpec with Matchers with Eventually {
                   commands))
               Single.error(new RuntimeException("expected failure"))
             }
+
```
fwiw `rxjava.components` has been deprecated so we _could_ omit multi-party submissions here. That said, given that you’ve already written the code we might as well keep it.

cocreature commented on language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/SubmitCommandsRequest.java:
```diff
@@ -33,16 +38,35 @@ public SubmitCommandsRequest(@NonNull String workflowId, @NonNull String applica
         this.applicationId = applicationId;
         this.commandId = commandId;
-        this.party = party;
+        this.party = actAs.get(0);
```
Maybe a good idea to throw a better error message than just an index out of bounds here?

cocreature reviewed (approved):
Nice work, thank you!

cocreature commented:
> Furthermore, I believe we should consider using a different approach next time we need to add something, this is becoming unwieldy. Maybe using a type wrapping command submission parameters coupled with a builder can be enough.

Fully agree. I was briefly discussing this with Gary earlier this week. We decided to go with this approach for now since it’s (more or less) backwards compatible but I think the API design needs a bit of work. There are two issues here imho:

1. As you already suggested, we should have a `Commands` type instead of accepting the parameters separately everywhere.
2. This type should imho be constructed via some type of builder pattern rather than overloads for every possible combination. Builders are also relatively easy wtr to backwards compatibility.

Not quite sure if we should try to shoehorn this into the existing API or create `CommandClient2` and the like which follow this pattern and only add new features to that.

stefanobaghino-da commented:
> Not quite sure if we should try to shoehorn this into the existing API or create `CommandClient2` and the like which follow this pattern and only add new features to that.

Would it be possible to shoehorn it in and deprecate existing methods? We can keep the existing methods for as long as we need.

cocreature commented:
> > 
> 
> Would it be possible to shoehorn it in and deprecate existing methods? We can keep the existing methods for as long as we need.

Yeah that could be an option.

garyverhaegen-da commented on language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/SubmitCommandsRequest.java:
```diff
@@ -76,14 +106,40 @@ public static SubmitCommandsRequest fromProto(CommandsOuterClass.Commands comman
                 .setApplicationId(applicationId)
                 .setCommandId(commandId)
-                .setParty(party)
+                .setParty(actAs.get(0))
```
Added
```
        if (actAs.size() == 0) {
            throw new IllegalArgumentException("actAs must have at least one element");
        }
```
in both places where we access `actAs.get(0)` (`toProto` and constructor).

garyverhaegen-da commented:
I think I've addressed all comments to the point where we can merge this.

@stefanobaghino-da if you feel strongly about removing the `list` function I'll be happy to do that in a subsequent PR on Monday.

Also would you both please mind taking a quick look at the [third commit](f753179)? I have no idea how important it is, but it seemed odd to leave it out. (Now that I noticed it; I hadn't so far. 🤷)

garyverhaegen-da commented:
Also happy to take a stab at drafting a better API next week, but I don't think that should be done in this PR.

cocreature commented:
third commit looks good to me.

stefanobaghino-da commented:
> @stefanobaghino-da if you feel strongly about removing the list function I'll be happy to do that in a subsequent PR on Monday.

If `list` makes it more readable please keep it in, but if I'm reading [the docs](https://docs.oracle.com/javase/8/docs/api/java/util/Arrays.html#asList-T...-) correctly I believe that `Arrays.asList` is equivalent in that regard:

> This method also provides a convenient way to create a fixed-size list initialized to contain several elements:

     List<String> stooges = Arrays.asList("Larry", "Moe", "Curly");

garyverhaegen-da commented:
> > @stefanobaghino-da if you feel strongly about removing the list function I'll be happy to do that in a subsequent PR on Monday.
> 
> If `list` makes it more readable please keep it in, but if I'm reading [the docs](https://docs.oracle.com/javase/8/docs/api/java/util/Arrays.html#asList-T...-) correctly I believe that `Arrays.asList` is equivalent in that regard:
> 
> > This method also provides a convenient way to create a fixed-size list initialized to contain several elements:
> 
> ```
>  List<String> stooges = Arrays.asList("Larry", "Moe", "Curly");
> ```

Right. My brain is not working today. I've updated to use `Arrays.asList`.

stefanobaghino-da commented:
> Right. My brain is not working today. I've updated to use `Arrays.asList`.

Lucky you, my brain doesn't work most days. 😆
